src/main.rs

```diff
@@ -0,0 +1,272 @@
+        std::fs::File::open(std::env::var("SSL_CERT_PATH").expect("SSL_CERT_PATH missing"))
+            .expect("read ssl cert"),
+    );
+    let mut key_file = std::io::BufReader::new(
+        std::fs::File::open(std::env::var("SSL_KEY_PATH").expect("SSL_KEY_PATH missing"))
+            .expect("read ssl key"),
+    );
+
+    let tls_certs = rustls_pemfile::certs(&mut certs_file)
+        .collect::<Result<Vec<_>, _>>()
+        .expect("failed to parse cert pem");
+
+    let tls_key = rustls_pemfile::rsa_private_keys(&mut key_file)
+        .next()
+        .expect("no keys found")
+        .expect("invalid key found");
+
+    // set up TLS config options
+    let tls_config = rustls::ServerConfig::builder()
+        .with_no_client_auth()
+        .with_single_cert(tls_certs, rustls::pki_types::PrivateKeyDer::Pkcs1(tls_key))
+        .expect("invalid tls config");
+
+    let Ok(server) = HttpServer::new(move || {
+        App::new()
+            .route("/", web::get().to(index))
+            .route("/token/login", web::post().to(login_token))
+            .route("/token/refresh", web::get().to(refresh_token))
+            .route("/reboot", web::post().to(reboot))
+            .route("/reload-network", web::post().to(reload_network))
+            .service(
+                Files::new(
+                    "/static",
+                    std::fs::canonicalize("static").expect("static folder not found"),
+                )
+                .show_files_listing(),
+            )
+    })
+    .bind_rustls_0_22("0.0.0.0:1977", tls_config) else {
+        error!("cannot bind server");
+        return;
+    };
+
+    let mut centrifugo =
+        Command::new(std::fs::canonicalize("centrifugo").expect("centrifugo not found"))
+            .spawn()
+            .expect("Failed to spawn child process");
+
+    tokio::select! {
+        _ = server.run() => {debug!("1");centrifugo.kill().await.expect("kill failed")},
+        _ = centrifugo.wait() => {}
+    }
+}
+
+async fn index() -> actix_web::Result<NamedFile> {
+    debug!("index() called");
+
+    // trigger omnect-device-service to republish
+    let res = post("/republish/v1", None).await;
+
+    if let Some(e) = res.error() {
+        error!("republish failed: {e}");
+
+        return Err(actix_web::error::ErrorInternalServerError(
+            "republish failed",
+        ));
+    }
+
+    Ok(NamedFile::open(
+        std::fs::canonicalize("static/index.html").expect("static/index.html not found"),
+    )?)
+}
+
+async fn login_token(auth: BasicAuth) -> impl Responder {
+    debug!("login_token() called");
+    let Ok(user) = std::env::var("LOGIN_USER") else {
+        error!("login_token: missing user");
+        return HttpResponse::build(StatusCode::INTERNAL_SERVER_ERROR).finish();
+    };
+    let Ok(password) = std::env::var("LOGIN_PASSWORD") else {
+        error!("login_token: missing password");
+        return HttpResponse::build(StatusCode::INTERNAL_SERVER_ERROR).finish();
+    };
```

Comment:
Is it sensible to have a running server instance if there are missing configuration properties, after all? I'd probably move this to main and make these to mandatory attributes of the server instance to ensure that these properties are always set for a valid server.

Reply:
it is only required for login route